File: Backend/app/schemas/household.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class UserType(str, Enum):
    family = "family"
    asha = "asha"
    anganwadi = "anganwadi"
    health_worker = "health_worker"


# ─── Request schemas ─────────────────────────────────────────────────────────


class HouseholdCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    primary_language: str = Field(default="en", max_length=10)
    user_type: UserType = Field(default=UserType.family)
    username: str = Field(..., max_length=50)
    password: str = Field(..., min_length=6, max_length=100)
    village_town: str | None = Field(default=None, max_length=200)
    state: str | None = Field(default=None, max_length=100)
    district: str | None = Field(default=None, max_length=100)


class HouseholdUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=200)
    primary_language: str | None = Field(default=None, max_length=10)
    user_type: UserType | None = Field(default=None)
    village_town: str | None = None
    state: str | None = None
    district: str | None = None
    preferences: dict | None = None
    last_onboarded_at: datetime | None = None


class HouseholdPreferences(BaseModel):
    ai_tone: str = Field(default="simple")
    language: str = Field(default="en")  # en, hi, mr
    voice_mode: str = Field(default="regional")
    health_focus: str = Field(default="general")


# ─── Response schemas ─────────────────────────────────────────────────────────


class HouseholdResponse(BaseModel):
    id: str
    username: str | None
    name: str
    primary_language: str
    user_type: UserType
    village_town: str | None
    state: str | None
    district: str | None
    preferences: dict
    last_onboarded_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: Mobile/lib/features/households/data/models/household.dart
import 'package:json_annotation/json_annotation.dart';

part 'household.g.dart';

enum UserType {
  family,
  asha,
  anganwadi,
  @JsonValue('health_worker')
  healthWorker,
}

@JsonSerializable(fieldRename: FieldRename.snake, explicitToJson: true)
class HouseholdPreferences {
  const HouseholdPreferences({
    this.aiTone = 'simple',
    this.language = 'en',
    this.voiceMode = 'regional',
    this.healthFocus = 'general',
  });

  final String aiTone;
  final String language;
  final String voiceMode;
  final String healthFocus;

  factory HouseholdPreferences.fromJson(Map<String, dynamic> json) =>
      _$HouseholdPreferencesFromJson(json);

  Map<String, dynamic> toJson() => _$HouseholdPreferencesToJson(this);
}

@JsonSerializable(fieldRename: FieldRename.snake, explicitToJson: true)
class Household {
  const Household({
    required this.id,
    required this.name,
    required this.primaryLanguage,
    required this.userType,
    required this.preferences,
    required this.createdAt,
    required this.updatedAt,
    this.username,
    this.villageTown,
    this.state,
    this.district,
    this.lastOnboardedAt,
  });

  final String id;
  final String? username;
  final String name;
  final String primaryLanguage;
  final UserType userType;
  final String? villageTown;
  final String? state;
  final String? district;
  final HouseholdPreferences preferences;
  final DateTime? lastOnboardedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory Household.fromJson(Map<String, dynamic> json) => _$HouseholdFromJson(json);

  Map<String, dynamic> toJson() => _$HouseholdToJson(this);
}

